<a href="https://colab.research.google.com/github/filchyboy/DS-Unit-2-Kaggle-Challenge/blob/master/DS_Sprint_Challenge_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science, Unit 2_
 
# Sprint Challenge: Predict Steph Curry's shots 🏀

For your Sprint Challenge, you'll use a dataset with all Steph Curry's NBA field goal attempts. (Regular season and playoff games, from October 28, 2009, through June 5, 2019.) 

You'll predict whether each shot was made, using information about the shot and the game. This is hard to predict! Try to get above 60% accuracy. The dataset was collected with the [nba_api](https://github.com/swar/nba_api) Python library.

In [0]:
%%capture
!pip install category_encoders

In [0]:
import datetime
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.pipeline import make_pipeline
import category_encoders as ce
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import StandardScaler

In [0]:
%%capture
import sys

if 'google.colab' in sys.modules:
    # Install packages in Colab
    !pip install category_encoders==2.*
    !pip install pandas-profiling==2.*

In [0]:
# Read data
import pandas as pd
url = 'https://drive.google.com/uc?export=download&id=1fL7KPyxgGYfQDsuJoBWHIWwCAf-HTFpX'
df = pd.read_csv(url)

# Check data shape
assert df.shape == (13958, 20)

In [5]:
df.shape

(13958, 20)

In [6]:
df.describe()

game_id  game_event_id  ...  shot_made_flag  scoremargin_before_shot
count  1.395800e+04   13958.000000  ...    13958.000000             13958.000000
mean   2.442837e+07     270.438458  ...        0.472919                 1.615561
std    7.226620e+06     169.927170  ...        0.499284                10.127139
min    2.090002e+07       2.000000  ...        0.000000               -39.000000
25%    2.120091e+07     109.000000  ...        0.000000                -4.000000
50%    2.150026e+07     275.500000  ...        0.000000                 1.000000
75%    2.170096e+07     398.000000  ...        1.000000                 8.000000
max    4.180040e+07     752.000000  ...        1.000000                43.000000

[8 rows x 10 columns]

To demonstrate mastery on your Sprint Challenge, do all the required, numbered instructions in this notebook.

To earn a score of "3", also do all the stretch goals.

You are permitted and encouraged to do as much data exploration as you want.

**1. Begin with baselines for classification.** Your target to predict is `shot_made_flag`. What is your baseline accuracy, if you guessed the majority class for every prediction?

**2. Hold out your test set.** Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

**3. Engineer new feature.** Engineer at least **1** new feature, from this list, or your own idea.
- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
- **Opponent**: Who is the other team playing the Golden State Warriors?
- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
- **Made previous shot**: Was Steph Curry's previous shot successful?

**4. Decide how to validate** your model. Choose one of the following options. Any of these options are good. You are not graded on which you choose.
- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
- **Train/validate/test split: random 80/20%** train/validate split.
- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

**5.** Use a scikit-learn **pipeline** to **encode categoricals** and fit a **Decision Tree** or **Random Forest** model.

**6.** Get your model's **validation accuracy.** (Multiple times if you try multiple iterations.) 

**7.** Get your model's **test accuracy.** (One time, at the end.)


**8.** Given a **confusion matrix** for a hypothetical binary classification model, **calculate accuracy, precision, and recall.**

### Stretch Goals
- Engineer 4+ new features total, either from the list above, or your own ideas.
- Make 2+ visualizations to explore relationships between features and target.
- Optimize 3+ hyperparameters by trying 10+ "candidates" (possible combinations of hyperparameters). You can use `RandomizedSearchCV` or do it manually.
- Get and plot your model's feature importances.



## 1. Begin with baselines for classification. 

>Your target to predict is `shot_made_flag`. What would your baseline accuracy be, if you guessed the majority class for every prediction?

In [7]:
df['shot_made_flag'].mean()

0.4729187562688064

Shot baseline is 47%. This can be found in the df describe I did above and is reflected as well in this specific call here: 0.4729187562688064

## 2. Hold out your test set.

>Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

In [8]:
df.columns

Index(['game_id', 'game_event_id', 'player_name', 'period',
       'minutes_remaining', 'seconds_remaining', 'action_type', 'shot_type',
       'shot_zone_basic', 'shot_zone_area', 'shot_zone_range', 'shot_distance',
       'loc_x', 'loc_y', 'shot_made_flag', 'game_date', 'htm', 'vtm',
       'season_type', 'scoremargin_before_shot'],
      dtype='object')

Found the season column in 'game_date'

In [0]:
df['game_date']= pd.to_datetime(df['game_date']) 
df['season_year'] = pd.DatetimeIndex(df['game_date']).year
df['season_month'] = pd.DatetimeIndex(df['game_date']).month

Converted 'game_date' to datetime pandas and then pulled out the year and added the feature of season with 'season_year'

In [0]:
mask = (df['game_date'] > pd.to_datetime('2018-10-01')) & (df['game_date'] <= pd.to_datetime('2019-06-30'))
test = df.loc[mask]

In [11]:
test.shape

(1709, 22)

Created a mask against the original 'game_date' feature which extended from Oct 2018 to June 2019 and then ran its shape to confirm accuracy.

## 3. Engineer new feature.

>Engineer at least **1** new feature, from this list, or your own idea.
>
>- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
>- **Opponent**: Who is the other team playing the Golden State Warriors?
>- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
>- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
>- **Made previous shot**: Was Steph Curry's previous shot successful?

    

In [12]:
df['season_year'].head(),df['season_month'].head()

(0    2009
 1    2009
 2    2009
 3    2009
 4    2009
 Name: season_year, dtype: int64, 0    10
 1    10
 2    10
 3    10
 4    10
 Name: season_month, dtype: int64)

I already created 2 new features 'season_year' & 'season_month'

In [13]:
teams = np.array(df['vtm'])
teams

array(['HOU', 'HOU', 'HOU', ..., 'TOR', 'TOR', 'TOR'], dtype=object)

I could create a dictionary to engineer a feature using english names for the team abbreviations. But I have already engineered two features and I will move on for now.


## **4. Decide how to validate** your model. 

>Choose one of the following options. Any of these options are good. You are not graded on which you choose.
>
>- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
>- **Train/validate/test split: random 80/20%** train/validate split.
>- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

In [0]:
mask_train = (df['game_date'] > pd.to_datetime('2009-10-01')) & (df['game_date'] <= pd.to_datetime('2017-06-30'))
train = df.loc[mask_train]
mask_val = (df['game_date'] > pd.to_datetime('2017-10-01')) & (df['game_date'] <= pd.to_datetime('2018-06-30'))
val = df.loc[mask_val]

In [15]:
train.shape, val.shape, test.shape

((11081, 22), (1168, 22), (1709, 22))

In [16]:
df.isnull().sum()

game_id                    0
game_event_id              0
player_name                0
period                     0
minutes_remaining          0
seconds_remaining          0
action_type                0
shot_type                  0
shot_zone_basic            0
shot_zone_area             0
shot_zone_range            0
shot_distance              0
loc_x                      0
loc_y                      0
shot_made_flag             0
game_date                  0
htm                        0
vtm                        0
season_type                0
scoremargin_before_shot    0
season_year                0
season_month               0
dtype: int64

In [0]:
def wrangle(X):
    """Wrangle train, validate, and test sets in the same way"""
    
    # Drop columns
    none = ['game_id', 'game_event_id', 'loc_x', 'loc_y', 'player_name' ]
    X = X.drop(columns=none)
    # return the wrangled dataframe
    return X

In [0]:
train = wrangle(train)
val = wrangle(val)
test = wrangle(test)

In [0]:
target = 'shot_made_flag'
features = train.columns.drop(target)
X_train = train[features]
y_train = train[target]
X_val = val[features]
y_val = val[features]
X_test = test[features]

In [20]:
features.shape, X_train.shape, y_train.shape, X_val.shape, y_val.shape, X_test.shape

((16,), (11081, 16), (11081,), (1168, 16), (1168, 16), (1709, 16))

## 5. Use a scikit-learn pipeline to encode categoricals and fit a Decision Tree or Random Forest model.

In [0]:
pipeline = make_pipeline(
    ce.OrdinalEncoder(), 
    SimpleImputer(strategy='mean'), 
    RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1)
)

In [0]:
dt = make_pipeline(
    ce.OneHotEncoder(),
    SimpleImputer(), 
    DecisionTreeClassifier(max_depth=16, random_state=42)
)

In [41]:
dt.fit(X_train, y_train)
print('Decision Tree:')
print('Train Accuracy', dt.score(X_train, y_train))
print('Validation Accuracy', dt.score(X_val, y_val))

TypeError: ignored

In [22]:
X_val.columns

Index(['period', 'minutes_remaining', 'seconds_remaining', 'action_type',
       'shot_type', 'shot_zone_basic', 'shot_zone_area', 'shot_zone_range',
       'shot_distance', 'game_date', 'htm', 'vtm', 'season_type',
       'scoremargin_before_shot', 'season_year', 'season_month'],
      dtype='object')

In [23]:
X_train.columns

Index(['period', 'minutes_remaining', 'seconds_remaining', 'action_type',
       'shot_type', 'shot_zone_basic', 'shot_zone_area', 'shot_zone_range',
       'shot_distance', 'game_date', 'htm', 'vtm', 'season_type',
       'scoremargin_before_shot', 'season_year', 'season_month'],
      dtype='object')

In [39]:
pipeline.fit(X_train, y_train)
print('Train Accuracy', pipeline.score(X_train, y_train))
print('Validation Accuracy', pipeline.score(X_val, y_val))

TypeError: ignored

## 6.Get your model's validation accuracy

> (Multiple times if you try multiple iterations.)

In [40]:


pipeline = make_pipeline(
    ce.TargetEncoder(min_samples_leaf=1, smoothing=1), 
    SimpleImputer(strategy='median'), 
    RandomForestRegressor(n_estimators=100, n_jobs=-1, random_state=42)
)

k = 3
scores = cross_val_score(pipeline, X_train, y_train, cv=k, 
                         scoring='neg_mean_absolute_error')
print(f'MAE for {k} folds:', -scores)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
TypeError: invalid type promotion

  FitFailedWarning)


MAE for 3 folds: [nan nan nan]


## 7. Get your model's test accuracy

> (One time, at the end.)

## 8. Given a confusion matrix, calculate accuracy, precision, and recall.

Imagine this is the confusion matrix for a binary classification model. Use the confusion matrix to calculate the model's accuracy, precision, and recall.

<table>
  <tr>
    <td colspan="2" rowspan="2"></td>
    <td colspan="2">Predicted</td>
  </tr>
  <tr>
    <td>Negative</td>
    <td>Positive</td>
  </tr>
  <tr>
    <td rowspan="2">Actual</td>
    <td>Negative</td>
    <td style="border: solid">85</td>
    <td style="border: solid">58</td>
  </tr>
  <tr>
    <td>Positive</td>
    <td style="border: solid">8</td>
    <td style="border: solid"> 36</td>
  </tr>
</table>

### Calculate accuracy 

In [0]:
(85+36)/(85+36+58+8)

### Calculate precision

In [0]:
36/(58+36)

### Calculate recall

In [0]:
36/(8+36)